drivers/display/sh1106.zig

```diff
@@ -0,0 +1,371 @@
+            try self.execute_command(0xDB, &.{level});
+        }
+
+        pub fn nop(self: Self) !void {
+            try self.execute_command(0xE3, &.{});
+        }
+
+        pub fn charge_pump_setting(self: Self, enable: bool) !void {
+            const arg: u8 = if (enable) 0x8B else 0x8A;
+            try self.execute_command(0xAD, &.{arg});
+        }
+
+        /// Set output voltage (Vpp) of internal charge pump.
+        pub fn set_pump_voltage(self: Self, voltage: enum(u2) {
+            @"6.4",
+            @"7.4",
+            @"8.0",
+            @"9.0",
+        }) !void {
+            const cmd: u8 = 0x30 | @as(u8, @intFromEnum(voltage));
+            try self.execute_command(cmd, &.{});
+        }
+
+        const i2c_command_preamble: []const u8 = &.{I2C_ControlByte.command};
+        const i2c_data_preamble: []const u8 = &.{I2C_ControlByte.data_stream};
+
+        fn execute_command(self: Self, cmd: u8, argv: []const u8) !void {
+            try self.set_dc_pin(.command);
+
+            try self.dd.connect();
+            defer self.dd.disconnect();
+
+            const command_preamble: []const u8 = switch (options.mode) {
+                .spi_3wire, .spi_4wire => "",
+                .i2c => i2c_command_preamble,
+                .dynamic => switch (self.mode) {
+                    .i2c => i2c_command_preamble,
+                    .spi_3wire, .spi_4wire => "",
+                    .dynamic => unreachable,
+                },
+            };
+
+            try self.dd.writev(&.{ command_preamble, &.{cmd}, argv });
+        }
+
+        fn write_data(self: Self, data: []const u8) !void {
+            try self.set_dc_pin(.data);
+
+            try self.dd.connect();
+            defer self.dd.disconnect();
+
+            const data_preamble: []const u8 = switch (options.mode) {
+                .spi_3wire, .spi_4wire => "",
+                .i2c => i2c_data_preamble,
+                .dynamic => switch (self.mode) {
+                    .i2c => i2c_data_preamble,
+                    .spi_3wire, .spi_4wire => "",
+                    .dynamic => unreachable,
+                },
+            };
+
+            try self.dd.writev(&.{ data_preamble, data });
+        }
+
+        /// If present, sets the D/C pin to the required mode.
+        /// NOTE: This function must be called *before* activating the device
+        ///       via chip select, so before calling `dd.connect`!
+        fn set_dc_pin(self: Self, mode: enum { command, data }) !void {
+            switch (options.mode) {
+                .i2c, .spi_3wire => return,
+                .spi_4wire => {},
+                .dynamic => if (self.mode != .spi_4wire)
+                    return,
+            }
+
+            try self.dc_pin.write(switch (mode) {
+                .command => .low,
+                .data => .high,
+            });
+        }
+    };
+}
+
+pub const I2C_ControlByte = @import("ssd1106.zig");
```

Comment:
could put this into a common file?

Reply:
Sounds good - not sure where it should go though, because this type doesn't seem specific to display drivers. I won't overthink it for now and will put it in drivers/display/common.zig unless you have another preference.